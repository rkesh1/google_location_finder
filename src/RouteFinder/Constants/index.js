export const REACT_APP_BASE_URL="https://mock-api.dev.lalamove.com/";
export const API_CALL_ITERERATION = 3;
export const INTERNAL_SERVER_ERROR = "Internal Server Error";
export const ZOOM = 12;
export const IN_PROGRESS = "in progress";
export const FAILURE= "failure";
export const SUCCESS = "success";
