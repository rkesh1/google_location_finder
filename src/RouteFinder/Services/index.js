  /**
   * @method apiService
   * @description All api call use apiService method
   * @param {string} url api endPath
   * @param {Object} params parameter need to pass default blank object
   * @param {Object} data user data
   * @returns {Object} API response
   */
  const apiService = async (url, params = {}, data = {}) => {
    const defaultParams = {
      method: "get",
      headers: { "Content-type": "application/json" },
      data: data,
      ...params
    };
    return await fetch(url, defaultParams)
      .then(async response => await response.json())
      .catch((err) => err );
  };

  export default apiService;