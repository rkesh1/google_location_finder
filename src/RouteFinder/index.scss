  .container {
    display: flex;
    height: 100vh;
    @media only screen and (max-width: 1200px) {
      flex-direction: column;
    }

    & .leftContainer {
      width: 20%;
      border: 1px solid #ccc;
      @media only screen and (max-width: 1200px) {
        width: 100%;
      }
    }
  
    & .rightContainer {
      width: 80%;
      border: 1px solid #ccc;
      @media only screen and (max-width: 1200px) {
        width: 100%;
      }
    }
  }

  .loader-overlay {
    height: 100%;
    position: fixed;
    width: 100%;
    opacity: .5;
    background: #fff;
  }
  .loader-container {
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    left: 50%;
    z-index: 1;
    & .loader {
      border: 16px solid #f3f3f3; 
      border-top: 16px solid #3498db; 
      border-radius: 50%;
      width: 50px;
      height: 50px;
      animation: spin 2s linear infinite;
    }
    
    @keyframes spin {
      0% { transform: rotate(0deg); }
      100% { transform: rotate(360deg); }
    }
  }


  