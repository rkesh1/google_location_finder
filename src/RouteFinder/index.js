import React, {Fragment} from "react";
import SearchForm from "./component/SearchForm";
import GoogleMap from "./component/GoogleMap";
import apiService from "./Services/index";
import {API_CALL_ITERERATION, INTERNAL_SERVER_ERROR, ZOOM, SUCCESS, IN_PROGRESS, FAILURE, REACT_APP_BASE_URL} from './Constants/index'; 
import "./index.scss";


/**
 * A component App initilization
 * @returns {Element} Rendered App
 */
class RouteFinder extends React.Component {
  /**
   * @method constructor
   * @description By default sets the state
   * @param {object} props Incoming props
   */
  constructor(props) {
    super(props);
    this.state = {
      cordinateData: [],
      renderPath: false,
      showErr: false,
      errorResponse: "",
      loader: false,
      totalDistance: "",
      totalTime: "",
      clearTextField: false,
      map: null,
      resubmit: false,
    };
    this.iteration = 0;
  }

  componentDidMount() {
    const GOOGLE_API_KEY = process.env.REACT_APP_GOOGLE_API_KEY;
    const GoogleMAPURL = `https://maps.googleapis.com/maps/api/js?key=${GOOGLE_API_KEY}&libraries=places`;
    const scriptObj = document.createElement("script");
    scriptObj.type = "text/javascript";
    scriptObj.async = true;
    scriptObj.src = GoogleMAPURL;
    document.body.appendChild(scriptObj);

    scriptObj.addEventListener("load", () => {
      this.setState({ map: window.google.maps });
    });
  }

  /**
   * @method resetFormHandler
   * @description resetFormHandler use to clear form field
   */
  resetFormHandler = () => {
    this.setState({
      renderPath: false,
      clearTextField: true,
      showErr: false,
      resubmit: false,
    });
  };

  /**
   * @method resetFormCompleted
   * @description resetFormCompleted use for clear ref data
   */
  resetFormCompleted = () => {
    this.setState({
      clearTextField: false
    });
  };

      /**
   * @method clearField
   * @description clear field on event of cross button
   */
  clearMessage = (e) => {
    this.setState({
      renderPath: false,
      showErr: false,
    });
  }

  loaderInit = () => {
    this.setState({
      loader: true,
      showErr: false,
      renderPath: false,
    });
  }
  /**
   * @method getRouteOnToken
   * @description call api with token for get route path coordinate
   * @param {string} token geting from api
   */
  getRouteOnToken = async token => {
    this.iteration++;
    const url = `${REACT_APP_BASE_URL}route/${token}`;
    const responseData = await apiService(url);
    if (responseData.status === SUCCESS) {
      this.setState({
        cordinateData: responseData.path,
        renderPath: true,
        loader: false,
        totalDistance: responseData.total_distance,
        totalTime: responseData.total_time,
        resubmit: true
      });
    } else if (responseData.status === IN_PROGRESS && this.iteration < API_CALL_ITERERATION) {
      
      await this.getRouteOnToken(token);
    } else {
      const errorMsg =
          responseData.status === FAILURE
            ? responseData.error
            : INTERNAL_SERVER_ERROR;

      this.setState({
        showErr: true,
        errorResponse: errorMsg,
        renderPath: false,
        resubmit: true,
        loader: false,
      });
    }
  };

  /**
   * @method submitFormHandler
   * @description Submit form and and get tocken
   * @param {Object} data user form field data
   */
  submitFormHandler = async data => {
    const {source, destination} = data;
    this.iteration = 0;
    const url = `${REACT_APP_BASE_URL}route`;
    const params = {
      method: "post",
      data: { origin: source, destination: destination }
    };
    
    this.loaderInit();
    const responseData = await apiService(url, params);
    this.setState({resubmit: true});
    if (responseData.token) {
      this.getRouteOnToken(responseData.token); 
    }else{
    this.setState({
      errorResponse: INTERNAL_SERVER_ERROR,
      showErr: true,
      loader: false,
      renderPath: false
    });
    }
  };

  render() {
    const {
      cordinateData,
      renderPath,
      showErr,
      errorResponse,
      loader,
      totalDistance,
      totalTime,
      clearTextField,
      map,
      resubmit
    } = this.state;
    const message = {
      totalDistance,
      totalTime,
      renderPath,
      errorResponse,
      showErr
    };
    return (
      <div className="App">
        <div className="container">
          {loader && (
            <Fragment>
              <div className="loader-overlay">
              </div>
              <div className="loader-container">
                <div className="loader"></div>
              </div>
            </Fragment>
          )}
          <div className="leftContainer">
            {map && (
              <SearchForm
                submitFormHandler={this.submitFormHandler}
                message={message}
                resetFormHandler={this.resetFormHandler}
                clearTextField={clearTextField}
                resetFormCompleted={this.resetFormCompleted}
                resubmit={resubmit}
                clearMessage={this.clearMessage}
              />
            )}
          </div>
          <div className="rightContainer">
            {map && (
              <GoogleMap
                cordinateData={cordinateData}
                renderPath={renderPath}
                zoom={ZOOM}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
}
export default RouteFinder;
