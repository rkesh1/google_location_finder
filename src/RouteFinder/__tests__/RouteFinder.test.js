import React from "react";
import { shallow } from "enzyme";
import RouteFinder from "../index";
import  apiService from '../Services/index';
import {API_CALL_ITERERATION, IN_PROGRESS, FAILURE} from '../Constants/index';

describe("RouteFinder Component Test Case ", () => {
  let wrapper;

  /**
   * Renders a shallow version of the RouteFinder component
   */
  function renderShallow() {
    wrapper = shallow(<RouteFinder />);
  }

  it("RouteFinder Component render", () => {
    renderShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("Default map object is not available", () => {
    renderShallow();
    expect(wrapper.state().map).toBe(null);
  });

  
  it("Map Object is set", () => {
    renderShallow();
    wrapper.setState({
      map: {}
    });
    expect(wrapper.state().map).toBeDefined();
  });

  it("Fetch api test", () => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise); 
    renderShallow();
    expect(global.fetch).toHaveBeenCalledTimes(0);
  });

  it("API success for token call", () => {
    renderShallow();
    apiService('url');
    expect(global.fetch).toHaveBeenCalledTimes(1);
    global.fetch.mockClear(); // 7
  });

  it("For Fetch api call", async () => {
    const windowFetch = jest.spyOn(global, 'fetch');
    renderShallow();
    const mockData = {data: { token: 1}};
    const { data: { token } } = mockData;

    windowFetch.mockImplementation(() => Promise.resolve({
      token: 1,
      status: 200,
      json: () => Promise.resolve(mockData),
    }))

    const response = await apiService('url');
    
    expect(response.data.token).toBe(token)
    global.fetch.mockClear();

  });

  it("For Fetch failure", async () => {
    const windowFetch = jest.spyOn(global, 'fetch');
    renderShallow();
    const errorMessage = 'New Error'
    windowFetch.mockImplementation(() => Promise.reject(errorMessage))
    try {
      await apiService('url');
    }
    catch(e) {
      expect(e.message).toBe(errorMessage);
    }
    global.fetch.mockClear();
  });

  it("getRouteOnToken post call on success", async () => {
    const mockDirectionResponse = {
      status: "success",
      path: [
        ["22.47", "114.47"],
        ["22.48", "114.48"],
        ["22.49", "114.49"]
      ],
      total_distance: 20000,
      total_time: 1800
    };
    const {total_distance, total_time} = mockDirectionResponse;
    renderShallow();
    const instance = wrapper.instance();
    const windowFetch = jest.spyOn(global, 'fetch');
    windowFetch.mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockDirectionResponse),
    }));

    await instance.getRouteOnToken('token');
    
    expect(instance.state.totalDistance).toBe(total_distance);
    expect(instance.state.totalTime).toBe(total_time);
    global.fetch.mockClear();
  });

  it("getRouteOnToken post call on inprogress", async () => {
    const mockDirectionResponseRetry = {
      status: IN_PROGRESS
    };
    renderShallow();
    const instance = wrapper.instance();
    const windowFetch = jest.spyOn(global, 'fetch');
    windowFetch.mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockDirectionResponseRetry),
    }));
    await instance.getRouteOnToken('token');

    expect(instance.state.showErr).toBe(true);
    expect(global.fetch).toHaveBeenCalledTimes(API_CALL_ITERERATION);
    global.fetch.mockClear();
  });

  it("getRouteOnToken post call on failure", async () => {
    const mockDirectionFailure = {
      status: FAILURE
    };
    renderShallow();
    const instance = wrapper.instance();
    const windowFetch = jest.spyOn(global, 'fetch');
    windowFetch.mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockDirectionFailure),
    }));
    await instance.getRouteOnToken('token');
    expect(instance.state.showErr).toBe(true);
  });

  it("getRouteOnToken post call on error", async () => {
    const windowFetch = jest.spyOn(global, 'fetch');
    renderShallow();
    const instance = wrapper.instance();
    const errorMessage = 'New Error'
    windowFetch.mockImplementation(() => Promise.reject(errorMessage))
    try {
      await instance.getRouteOnToken('token');
    }
    catch(e) {
      expect(e.message).toBe(errorMessage);
    }
  });
});
