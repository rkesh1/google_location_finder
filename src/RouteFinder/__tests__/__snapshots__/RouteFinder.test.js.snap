// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteFinder Component Test Case  RouteFinder Component render 1`] = `ShallowWrapper {}`;
